tempesta_fw/http_parser.c

```diff
@@ -697,25 +698,28 @@ __FSM_STATE(RGen_HdrOtherV) {						\
 /* Process according RFC 7230 3.3.3 */
 #define TFW_HTTP_INIT_REQ_BODY_PARSING()				\
 __FSM_STATE(RGen_BodyInit) {						\
+	TfwStr *tbl = msg->h_tbl->tbl;					\
+									\
 	TFW_DBG3("parse request body: flags=%#x content_length=%lu\n",	\
 		 msg->flags, msg->content_length);			\
 									\
-	/*								\
-	 * TODO: If both "Transfer-Encoding:" and "Content-Length:"	\
-	 * headers are present,	then issue an error or remove		\
-	 * "Content-Length:" header.					\
-	 */								\
-	/*								\
-	 * TODO: If "Transfer-Encoding:" header is present,		\
-	 * then "chunked" coding must be the last in the list.		\
-	 */								\
-	if (msg->flags & TFW_HTTP_CHUNKED)				\
-		__FSM_MOVE_nofixup(RGen_BodyStart);			\
-	/*								\
-	 * TODO: If "Transfer-Encoding:" header is present and		\
-	 * there's NO "chunked" coding, then send 400 response		\
-	 * (Bad Request) and close the connection.			\
-	 */								\
+	if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_TRANSFER_ENCODING])) {	\
+		/* The alternative: remove "Content-Length" header. */	\
+		if (!TFW_STR_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH]))	\
+			return TFW_BLOCK;				\
+		if (msg->flags & TFW_HTTP_CHUNKED) {			\
+			/* "chunked" must be the last in the list. */	\
+			if (!(msg->flags & __TFW_HTTP_CHUNKED_LAST))	\
```

Comment:
This check is related to validation of `Transfer-Encoding` header. I think better place for it is `__parse_transfer_encoding()` function